### PR TITLE
Decouple the metrics from the internals of the metrics libraries [PLEN-87]

### PR DESCRIPTION
**Pull request by nicu-da**
The Metrics implementation now accepts the factories as dependencies, and this allows us to remove all the inheritance around the metric classes. This will also allow us to inject the factory implementation from canton thus using a single factory for the canton code and the sdk.

Furthermore, this cleanup will allow us to more easily change the metrics implementation in the future.

### CODE REVIEW

File: ledger/metrics/src/main/scala/com/daml/metrics/IndexerMetrics.scala [hunk 1]
```diff
@@ -22,7 +20,7 @@ class IndexerMetrics(val prefix: MetricName, override val registry: MetricRegist
     qualification = Debug,
   )
   val lastReceivedRecordTime: Gauge[Long] =
-    gauge(prefix :+ "last_received_record_time", 0)
+    factory.gauge(prefix :+ "last_received_record_time", 0L)(MetricsContext.Empty)
```
Comment (lucb-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/IndexerMetrics.scala:23]: We will have to add relevant labels after fully switching to OpenTelemetry.

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala [hunk 2]
```diff
@@ -93,15 +93,15 @@ private[platform] object InMemoryStateUpdater {
       InstrumentedExecutors.newWorkStealingExecutor(
         metrics.daml.lapi.threadpool.indexBypass.prepareUpdates,
         prepareUpdatesParallelism,
-        metrics.registry,
+        metrics.dropwizardFactory.registry,
```
Comment (lucb-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:96]: All this code is still dependent on Dropwizard.
Not sure how to resolve that. Maybe with the in-memory factory, but I don't know if it makes sense in this context.
Likely something to be recorded in a ticket somewhere.

**Review by lucb-da — COMMENTED** [review]
Good cleanup.
Some small comments.
There are is a few direct reference to Dropwizard.

File: ledger/metrics/src/main/scala/com/daml/metrics/IndexDBMetrics.scala [hunk 3]
```diff
@@ -57,7 +54,7 @@ class IndexDBMetrics(override val prefix: MetricName, override val registry: Met
   )
   val loadPartyEntries: DatabaseMetrics = createDbMetrics("load_party_entries")
 
-  object storeTransactionDbMetrics extends DatabaseMetrics(prefix, "store_ledger_entry", registry)
+  object storeTransactionDbMetrics extends DatabaseMetrics(prefix, "store_ledger_entry", factory)
```
[the remark this answers is not in the document]
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/IndexDBMetrics.scala:57]: For some reasons that's a pattern that was used quite a few times, I would say we should change it a separate PR but indeed you are correct

File: ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala (hunk 2 above)
Comment (nicu-da) [reply ledger/participant-integration-api/src/main/scala/platform/index/InMemoryStateUpdater.scala:96]: As part of the switch to opentelemetry this should also be solved.

File: ledger/metrics/src/main/scala/com/daml/metrics/IdentityProviderConfigStoreMetrics.scala [hunk 4]
```diff
@@ -3,23 +3,26 @@
...
 
-  val cacheByIssuer = new CacheMetrics(prefix :+ "cache_by_issuer", registry)
+  val cacheByIssuer = new CacheMetrics(prefix :+ "cache_by_issuer", factory)
   val createIdpConfig: DatabaseMetrics = createDbMetrics("create_identity_provider_config")
   val getIdpConfig: DatabaseMetrics = createDbMetrics("get_identity_provider_config")
   val deleteIdpConfig: DatabaseMetrics = createDbMetrics("delete_identity_provider_config")
   val updateIdpConfig: DatabaseMetrics = createDbMetrics("update_identity_provider_config")
   val listIdpConfigs: DatabaseMetrics = createDbMetrics("list_identity_provider_configs")
+
+  private def createDbMetrics(name: String) = {
```
[the remark this answers is not in the document]
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/IdentityProviderConfigStoreMetrics.scala]: Done

**Review by lucb-da — APPROVED** [review]
LGTM

**Pull request merged**